smrt/inputs/make_bedrock.py

```diff
@@ -0,0 +1,236 @@
+automatically load a specific bedrock model and provides some bedrock permittivity formulae as well.
+
+Example::
+
+    from smrt import make_bedrock
+    bedrock = make_bedrock(substrate_model="flat", permittivity_model="christianson", bedrock_type="unfrozen bedrock", temperature=273.15)
+
+It is recommended to first carefuly read the documentation of `make_bedrock` and then explore the different types of bedrock
+models.
+"""
+
+from numbers import Number
+from typing import Callable, Union, Any, Dict, Tuple, Optional
+
+import numpy as np
+
+from smrt.core.error import SMRTError
+from smrt.core.interface import get_substrate_model
+from smrt.core.globalconstants import PERMITTIVITY_OF_FREE_SPACE 
+from smrt.core.interface import Substrate
+
+# --- 1. DATA TABLES ---
+
+# Hartlieb et al. (2016) Data: {rock_type: (real_permittivity, imaginary_permittivity)}
+# frequency = 2450 MHz, temperature = 20 degc
+HARTLIEB_2016_DATA: Dict[str, Tuple[float, float]] = {
+    'granite': (5.45, 0.038),
+    'basalt': (7.67, 0.270),
+    'sandstone': (7.67, 0.081),
+}
+
+# # Tulaczyk and Foley (2020) Data: {material_type: (real_permittivity, conductivity)}
+# # Frequency ?, temperature 
+# TULACZYK_2020_DATA: Dict[str, Tuple[float, float]] = {
+#     'glacier ice': (3.2, 0.000078),
+#     'frozen bedrock': (2.7, 0.0002),
+#     'marine ice': (3.4, 0.00036),
+#     'saturated bedrock': (9.5, 0.0055), # Midpoints used
+#     'saline basal ice': (3.4, 0.02),    # Upper bound
+#     'sandy till': (13.0, 0.02),         # Midpoint & Upper bound
+#     'subglacial water': (88.0, 0.04),
+#     'fairbanks silt': (24.0, 0.043),
+#     'clay-bearing till': (13.0, 0.0575),# Midpoints used
+#     'clay': (31.0, 0.24),
+#     'marine clay': (31.0, 0.55),        # Midpoint used
+#     'seawater': (79.0, 2.90),
+#     'brine': (62.0, 4.89),
+# }
+
+# Christianson et al. (2016) Data: {material_type: (real_permittivity, conductivity)}
+# frequency = 500MHZ, temperature ? (depend of medium) 
+CHRISTIANSON_2016_DATA: Dict[str, Tuple[float, float]] = {
+    'glacier ice': (3.2, 7.0e-5),
+    'marine ice': (3.4, 5.7e-4),
+    'groundwater ice': (3.2, 6.6e-4),
+    'debris-laden ice': (3.1, 8.0e-5),
+    'seawater': (77.0, 2.9),
+    'freshwater': (80.0, 1.0e-4),
+    'groundwater': (80.0, 0.37),
+    'sand': (2.6, 1.3e-4),
+    'groundwater till': (36.0, 0.037),
+    'freshwater till': (13.0, 2.5e-4), # Midpoints used
+    'frozen till': (2.9, 3.4e-4),
+    'frozen bedrock': (2.7, 2.0e-4),
+    'unfrozen bedrock': (12.0, 0.0048),
+}
+
+# --- 2. PERMITIVITY FUNCTIONS  ---
+
+def constant_permittivity_Hartlieb_2016(bedrock_type: str) -> complex:
+    """
+    Returns the permittivity constant for granite, basalt, sandstone based on Hartlieb et al. (2016) http://dx.doi.org/10.1016/j.mineng.2015.11.008.
+    
+    This function uses a dictionary (HARTLIEB_2016_DATA) to look up values instead of if/elif chain.
```

Comment:
the documentation is for SMRT user (not for the developers), provide information that matters for the calculation (most important is the approximations and limitations)